### DISCUSSION TITLE
why do we have Vector, Size, Point, when they are actually the same?

### DISCUSSION THREAD

**Discussion by martinrhan** [AvaloniaUI/Avalonia#16278] [Q&A]
It is fine to just use a same struct type for all of them.

**Comment by stevemonaco** [comment]
They're conceptually different and support different operations, even if they hold two doubles as state.

Vector: (x, y) quantity that doesn't necessarily represent anything spatial. Has many additional math ops.
Point: (x, y) quantity that represents a location in 2D space
Size: (width, height) quantity that represents area in 2D space.

**Reply by martinrhan** [reply]
But they are usually used interchangely. For example, size is used as a vector or a point, point is used as a vector or a size. Given such fact, they should at least support some low cost casting, and it is better to be implicit.

**Reply by stevemonaco** [reply]
They shouldn't be used interchangeably. They exist because they represent different problem domain concepts and API. Named concepts have important contextual meaning.

I have my own implementations for a drawing app, so I'm juggling Vector/Point/Size/Rect types between my own, Avalonia's, System.Drawing's, and SkiaSharp's. I do use implicit conversions that haven't bit me (yet), but I don't think that's a good idea for a framework. You can write extension methods today to deal with the conversions if your main complaint is lack of conciseness.
